Yoast/duplicate-post
# P3-308 Prevent the original post from being listed in the link suggestions

## Context


* We want the Yoast SEO Premium link suggestions to give meaningful suggestions to users by preventing the original post of a Rewrite & Republish copy from being listed in the suggestions.

## Summary


This PR can be summarized in the following changelog entry:

* Fixes a bug where the original post of a Rewrite & Republish copy was listed in the Yoast SEO Premium link suggestions.

## Relevant technical choices:

- Uses the new filter `wpseo_link_suggestions_indexables` that was recently added in Premium. This filter runs very early, before the post is set and even before the WordPress GET request runs. The only way I could think of to get the current post is by using the `object_id` parameter of the link suggestions request. I'd like this part to be reviewed by an architect or senior dev.
- The new method is placed in the Block_Editor class even if it is used for both the Block Editor and Classic Editor. RIght now, this class is always instantiated and Duplicate Post doesn't have a more suitable class to be used. It's basically the same issue already tracked in [P3-296](https://yoast.atlassian.net/browse/P3-296) where there's a proposal for some refactoring to introduce a few "compatibility" classes. A class for Yoast SEO compatibility would be nice, but for now this is out of the scope of this PR.

## Test instructions

### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

**Reproduce the bug:**
- Activate Yoast SEO Premium and Duplicate Post (latest develop branch).
- Go to SEO > Tools and run the SEO data optimization.
- Test with the Block Editor.
- Pick up a published post that has some long content. Recommended: one of the auto-generated posts in the Docker environment. Alternatively, create a long-content post and publish it.
- Create a Rewrite & Republish copy of the above post.
- Open the Yoast sidebar by clicking the Yoast icon in the top bar.
- Open the "Internal linking suggestions" panel.
- If the SEO data optimization and link suggestions ran correctly, you should see a list of suggested posts to link to.
- Observe that the list of suggestions contains also the original post.
- Install and activate the Classic Editor plugin.
- Observe the same happens with the Classic Editor: the link suggestions are in the "Yoast internal linking" meta box in the sidebar.

**Test the fix:**
- Switch Duplicate Post to this branch and build: `composer install && yarn && grunt build:js`
- Deactivate the Classic Editor plugin.
- Repeat the steps above or just refresh the page of the Rewrite & Republish copy you created previously.
- Observe that the list of suggestions does not contain the original post any longer.
- Repeat the test using the Classic Editor.


### Test instructions for QA when the code is in the RC




QA can test this PR by following these steps:

**Reproduce the bug:**
- Activate Yoast SEO Premium and Duplicate Post latest released versions.
- Go to SEO > Tools and run the SEO data optimization.
- Test with the Block Editor.
- Pick up a published post that has some long content. Recommended: one of the auto-generated posts in the Docker environment. Alternatively, create a long-content post and publish it.
- Create a Rewrite & Republish copy of the above post.
- Open the Yoast sidebar by clicking the Yoast icon in the top bar.
- Open the "Internal linking suggestions" panel.
- If the SEO data optimization and link suggestions ran correctly, you should see a list of suggested posts to link to.
- Observe that the list of suggestions contains also the original post.
- Install and activate the Classic Editor plugin.
- Observe the same happens with the Classic Editor: the link suggestions are in the "Yoast internal linking" meta box in the sidebar.

**Test the fix:**
- Use the Duplicate Post Release Candidate that contains this fix.
- Deactivate the Classic Editor plugin.
- Repeat the steps above or just refresh the page of the Rewrite & Republish copy you created previously.
- Observe that the list of suggestions does not contain the original post any longer.
- Repeat the test using the Classic Editor.

## Impact check

This PR affects the following parts of the plugin, which may require extra testing:

*

## UI changes


## Documentation


## Quality assurance


Fixes [P3-308]
